[isaacchua0309] iP

isaacchua0309 opened the pull request:
# **Boblet - Your Friendly Task Manager** 📝

> *“Your mind is for having ideas, not holding them.”* – David Allen

Boblet helps you manage your tasks efficiently so you can focus on what truly matters. It's:

- **Text-based** 📄  
- **Easy to use** 🏆  
- **Super FAST** ⚡  

## **How to Use Boblet**
1. **Download** the latest version of Boblet from [[here](https://github.com/isaacchua0309/ip/releases](https://github.com/isaacchua0309/ip/releases).  
2. **Run the application** using:
   ```sh
   java -jar Boblet.jar
   ```
3. **Start managing your tasks** using simple commands!  
4. **Enjoy a more organized life!** 😃

## **Features**
- ✅ **Managing tasks**
- ⏳ **Setting deadlines**
- 📅 **Scheduling events**
- 🔍 **Finding tasks easily**
- 🗂️ **Saving and loading tasks automatically**

---

## **Code Example** 🖥️
If you are a Java developer, here’s a snippet of the **main method**:

```java
public class Main {
    public static void main(String[] args) {
        new Boblet("data/tasks.txt").run();
    }
}
```

## **To-Do List for Boblet Development** 📌
- [x] Implement basic task management  
- [x] Add deadline and event support  
- [x] Enable task searching  
- [ ] Implement recurring tasks  
- [x] Add a graphical user interface (GUI)  

---

### **Why Choose Boblet?**
✔️ **Simple & lightweight**  
✔️ **No unnecessary features**  
✔️ **Designed for efficiency**  

So, what are you waiting for? Try **Boblet** today! 🚀  

---

This PR follows **[GitHub](https://github.com/your-repo) Flavored Markdown (GFMD)** by incorporating:  
- ✅ **Headings** (#, ##, ###)  
- ✅ **Bullet Lists** (-, *)  
- ✅ **Numbered Lists** (1., 2., 3.)  
- ✅ **Fenced Code Blocks** (```java, ```sh)  
- ✅ **Task Lists** ([x], [ ])  
- ✅ **Emojis** (🚀, 📝, ✅)  
- ✅ **Blockquotes** (>)  
- ✅ **Hyperlinks** ([GitHub](https://github.com/isaacchua0309/ip/releases)  
- ✅ **Inline Code** (`java -jar Boblet.jar`)  
- ✅ **Text Formatting** (**bold**, *italic*, ~~strikethrough~~)  

🎉 *Looking forward to merging this PR into `main`!* 🚀

Commit:
Level-0 Tester

Commit:
Add Level-1 functionality: Greet, Echo, Exit

Commit:
Add Level-2 functionality: Add and List

Commit:
Add more personality and engagement to Boblet

Commit:
Add Level-3 functionality: Mark tasks as done with A-Classes

Commit:
Add Level-4 functionality: Support ToDos, Deadlines, and Events with A-Inheritance

Commit:
Add text UI teting component for Boblet

Commit:
Added ability to handle errors to Boblet

Commit:
Added Level-5 functionality: Error Handling to Boblet

Commit:
Add Level-6 functionality: Support Delete

Commit:
Added Enums component for Boblet

Commit:
Updated README file for Boblet

Commit:
Add Level-7 functionality: Implement task saving and loading with file handling

Commit:
Merge branch-Level-7 into master

Commit:
Add Level-8 functionality: Boblet date and time handling

Commit:
Merge branch-Level-8 into master

Commit:
Updated switch cases to include ShowDate command

Commit:
Added A-MoreOOP: Made the code more OOP and refactored closely related code as classes

Commit:
Added A-packages: Organised the classes into suitable java packages

Commit:
Added A-Gradle: Set up gradle to build and run Boblet

Commit:
Added A-JUnit functionality: Added JUnit tests to test behavior of code

Commit:
Added A-JavaDoc: Added JavaDoc comments to codebase

Commit:
Added Level-9 functionality: Allow user to search for keywords in task description

Commit:
Added A-CodingStandard: Standardised coding structure

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Added Level-9 functionality: Allow user to search for keywords in task description

Commit:
Add changes to Task for commit

Commit:
Edited Level 9 Branch to verify merging

Commit:
Added ip.jar to git ignore

Commit:
Level 9 merging testing

Commit:
CheckStyle: Implement and enforce style compliance

Boblet's codebase previously did not adhere strictly to a defined
style guide, leading to inconsistencies in formatting, import
ordering, and documentation. This made it harder to maintain
readability and ensure uniformity across the project.

To improve code quality and maintainability, let's:
- Integrate CheckStyle into the project
- Fix all violations, including import order, indentation, and Javadoc
- Ensure consistent spacing and formatting across all files

Following this commit, the project now fully complies with the
configured CheckStyle rules, ensuring a standardized and cleaner codebase.

Commit:
feat(ui): implement JavaFX GUI for Level 10

- Added JavaFX components to replace CLI-based interaction
- Integrated GUI with existing command execution logic
- Updated Main class to launch JavaFX application
- Followed SE-EDU coding standards for naming, layout, and comments

Closes #10

Commit:
refactor(ui): improve JavaFX UI responsiveness and stability

- Optimized layout resizing for better user experience
- Fixed event handling issues causing unresponsive UI
- Refactored JavaFX initialization to align with SE-EDU standards
- Improved error handling for invalid user inputs

Commit:
Merge branch 'branch-Level-10'

Commit:
build: update build.gradle dependencies and configurations

- Updated dependencies to latest stable versions
- Modified build settings for improved compatibility and performance
- Adjusted Gradle tasks to streamline the build process

Commit:
feat(ui): complete Level 10 implementation with JavaFX

- Fully implemented JavaFX-based GUI for user interactions
- Integrated GUI with backend command execution logic
- Improved UI responsiveness and event handling
- Refactored code to align with SE-EDU coding standards
- Finalized UI elements and layout for a polished user experience

Closes #10

Commit:
docs: enhance PR description with GitHub Flavored Markdown (GFMD)

- Added a structured heading for clarity
- Included a bullet list and a numbered list for better readability
- Incorporated a fenced code block with syntax highlighting
- Added a task list to track progress
- Used an emoji, blockquote, and hyperlink for improved formatting
- Applied inline code and text styling (bold, italic, strikethrough) for emphasis

This ensures familiarity with GFMD syntax and enhances PR readability.

chenxy12345 commented on src/test/java/boblet/util/ParserTest.java:
```diff
@@ -0,0 +1,199 @@
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import boblet.command.AddCommand;
```
I like how your import statements follow a consistent order among all classes

chenxy12345 commented on src/main/java/boblet/task/Event.java:
```diff
@@ -0,0 +1,100 @@
+ */
+public class Event extends Task {
+    private LocalDateTime at;
+
```
Maybe this variable name could be changed to something clearer such as dateTime ?

chenxy12345 commented on src/main/java/boblet/task/Event.java:
```diff
@@ -0,0 +1,100 @@
+     *
+     * @return A formatted string representation of the event date/time.
+     */
+    public String getAt() {
```
In relation to my previous comment, maybe this could be clearer by changing it to something like getDateTime ?

chenxy12345 commented on src/test/java/boblet/task/DeadlineTest.java:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    @Test
+    void testParseDateTimeWithMultipleFormats() {
```
Maybe this could be more descriptive ? eg. parseDateTimeWithMultipleFormats_success()

Commit:
Add assertions to enforce critical assumptions in Boblet

To improve code reliability and prevent unintended behavior, assertions
have been added to enforce assumptions about critical states in the
application.

The following assertions were added:
- Ensure that the  in  is never null or empty.
- Validate that  and  are always initialized.
- Confirm that user input in  is never null or empty.
- Ensure that parsed  objects are never null before execution.

These assertions will help in detecting unexpected states during
development, making debugging easier.

Commit:
Refactor code for improved readability and maintainability

Refactored various components of the Boblet application to enhance code quality, readability, and maintainability.

Changes include:
- Improved encapsulation by making  final in Boblet.
- Strengthened defensive programming with assertions to prevent null references.
- Standardized method behaviors to ensure consistency in parsing and execution.
- Enhanced task management logic by validating indices and input formats.

These improvements make the codebase more robust and maintainable while adhering to best practices for clean code.

Commit:
Refactor ListCommand to use Java Streams while maintaining assertions

Previously, the ListCommand used a traditional for-loop with a StringBuilder to construct the formatted task list. This approach, while functional, made the code more verbose and harder to read.

This commit improves the code quality by:
- Using Java Streams to format the task list in a more concise and readable way.
- Utilizing 'IntStream.range()' to iterate over task indices instead of a for-loop.
- Ensuring assertions remain in place to catch potential issues.
- Using '.mapToObj()' to construct task entries instead of manual 'StringBuilder.append()'.
- Applying '.reduce()' to concatenate task entries efficiently.

This refactor improves maintainability and makes the task listing code more declarative.

isaacchua0309 committed:
Merge pull request #1 from isaacchua0309/branch-A-Assertions

A-Assertions

isaacchua0309 committed:
Merge pull request #2 from isaacchua0309/branch-A-CodeQuality

A-CodeQuality

isaacchua0309 committed:
Merge pull request #3 from isaacchua0309/branch-A-Streams

A-Streams

Commit:
Tag previous commit as A-Assertions

The A-Assertions tag has been added to a previous commit that introduced Java Assertions into the codebase.

### Why this change?
Previously, task list operations relied on traditional loops for filtering and formatting, making the code more verbose. By using Streams, the operations are now more concise and declarative.

### What has been done?
- Refactored ListCommand to use Java Streams for formatting task lists.
- Ensured that Streams were used while maintaining all necessary assertions.
- Improved code clarity by removing redundant loops.

### Why this approach?
Using Streams allows for better functional programming practices, making the code easier to read and maintain. This update aligns with the Duke A-Streams requirement.

Commit:
fix: Prevent duplicate task addition and resolve logic error

Previously, the AddCommand class contained a bug where tasks were being added twice—once before entering the try-catch block and again inside it. This caused the duplicate check in TaskList to trigger an exception prematurely, preventing proper task addition after deletions.

To fix this:
- Removed the redundant tasks.addTask(task); inside the try-catch block.
- Ensured that task addition happens only once before saving to storage.
- Improved exception handling to properly return duplicate error messages when necessary.

With this fix, the application now correctly prevents duplicate tasks while allowing new additions after deletion.

isaacchua0309 committed:
Merge pull request #4 from isaacchua0309/branch-C-DetectDuplicates

C-DetectDuplicates

Commit:
Fix code styling and resolve Checkstyle issues

Addressed styling issues and ensured compliance with Checkstyle rules.

Changes include:
- Fixed FallThrough Checkstyle warning by adding necessary break statements in switch cases.
- Reorganized imports to avoid '.*' usage and match recommended import patterns.
- Corrected minor code formatting issues to adhere to style guidelines.

These changes ensure that the code is properly formatted and passes Checkstyle validation.

Commit:
feat(ui): Improve GUI for better user experience

The application's GUI has been improved to make it more user-friendly and visually appealing. Changes include:

- Enhanced input text box with rounded corners, color adjustments, and padding for better readability.
- Added a Help button for easy access to available commands.
- Introduced light and dark mode toggle for better accessibility.
- Adjusted dialog box layout for cleaner message display.
- Applied color schemes to differentiate user and bot messages.
- Fixed button overlapping and alignment issues.

These improvements create a more intuitive and visually appealing interface for users.

Commit:
fix: reorder imports to comply with Checkstyle rules

- Moved static imports to the top
- Reorganized imports to follow CustomImportOrder requirements
- Grouped standard Java, third-party, and static imports correctly

Commit:
docs: update README with latest features and user guide

- Added new features such as task search and date filtering
- Improved command descriptions and usage examples
- Included setup instructions and example interactions
- Updated architecture and future enhancements section
- Reformatted README to align with GitHub Markdown standards

Commit:
docs: add UI screenshot for product website

Added Ui.png to the docs folder to showcase the full GUI window for the product website.

isaacchua0309 committed:
Update README.md
